### PR TITLE
Acceptance tests for process handling: CFE-3025

### PR DESCRIPTION
**Pull request by tdlee**
The acceptance tests had not been covering 'tty' or 'process_owner'.  The two commits here begin to fill that gap.

The 'process_owner' test is standalone.

The 'tty' test has an associated source code change.  See the comment within its commit message for more details.

See issues CFE-3025, which itself emerged from CFE-3020 and CFE-1655

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @vpodzime can review this?

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2264/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2264/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2264/

**Comment by vpodzime** [comment]
Thanks for working on this @tdlee! One of the tests fails, unfortunately:
```
Error Message
FAIL (UNEXPECTED FAILURE) ./05_processes/01_matching/tty.cf

Stacktrace
2019-03-30T17:55:30+0000     info: Created directory "/home/jenkins/workspace/testing-pr/label/PACKAGES_HUB_x86_64_linux_redhat_7/cfengine-3.14.0a.cf8be3fbc/tests/acceptance/workdir/__05_processes_01_matching_tty_cf/tmp/TESTDIR.cfengine/."
R: /home/jenkins/workspace/testing-pr/label/PACKAGES_HUB_x86_64_linux_redhat_7/cfengine-3.14.0a.cf8be3fbc/tests/acceptance/./05_processes/01_matching/tty.cf FAIL
```

**Comment by tdlee** [comment]
Alas, the error message doesn't help me.  The test passed for me (RHEL6) otherwise I wouldn't have submitted it.

Bear in mind that 'tty' is advertised functionality but it lacks an acceptance test, and the functionality itself had, worryingly, remained untested.  So the intention behind this test is to fix two weaknesses:

- Advertised functionality that didn't seem to behave as advertised;

- Provision of a test for code that had not originally had tests associated with it.

In other words, this ''tty' test is designed to detect pre-existing weaknesses.

Indeed, there could be failure on other platforms.  But that's actually the point.  The intention is to verify that the core code is correct.  But hasn't it been non-functioning all along on most platforms?

Could I draw your attention to the detailed notes I made in the relevant commit (f8c44f2)?

Of course, there is the possibility that my test itself isn't correctly capturing the documentation's claims.  In that case, my test would need to be fixed.  Could you be a "second pair of eyes", showing me where the test doesn't match the documentation?

In short: Is the bug in the test?  Or is the bug a long-term lurker in the highly platform-varying code itself?

How do we proceed, please?

Thanks for your attention to this.

**Comment by vpodzime** [comment]
@tdlee looks like there is something weird going on when the tests run. Somehow when the *tty_match* test runs, *cf-agent* is spawning `ps` without the new output column you are adding in the same commit. When I do `make clean && make` and run the test again, it passes. I have no idea what's going on there, how it can compile things without your change first.

**Comment by tdlee** [comment]
Thanks for the reply. With regard to the tty components, I believe that:

- there is a source code bug in "systype.c" for several platforms;
- the lack of previous acceptance tests has allowed this bug to linger undetected for a long time;
- the new "tty.cf" test (alone) exposes the old source code bug;
- the "systype.c" aspect of my commit fixes it for Linux (with it, "tty.cf" should pass on Linux);
- the "systype.c" aspect also demonstrates the model for equivalent fixes on other susceptible platforms.

The previous code had lacked the request for the relevant column "ps" on Linux (and other platforms). That lack from "systype.c" is itself the bug.

In summary, the following is expected for Linux:

- test alone (with previous, unfixed "systype.c"): test should fail 
- test and "systype.c" bugfix together: test should pass

The new "tty.cf" here will probably expose equivalent failures (i.e. demonstrate the existing source code bug) on other platforms.  Equivalent "systype.c" adjustments would be needed there to fix such platforms.

I hope that helps.

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2289/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2289/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2289/

**Comment by tdlee** [comment]
Could someone provide an update, please, on why this seems stalled?  Thanks.

**Comment by vpodzime** [comment]
@cf-bottom jenkins with exotics, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2358/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2358/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2358/

**Comment by tdlee** [comment]
But do you have any idea, please, what is wrong?  The test passes in my environments.  I have outlined above what the test is trying to do.   I believe that a deeper bug may remain: the purpose of my test is to help expose that deeper bug.

Could you contact me, please?  Thanks.

**Comment by vpodzime** [comment]
> Could you contact me, please? Thanks.

It's still the same issue as I described in https://github.com/cfengine/core/pull/3582#issuecomment-480239343 above. Somehow, the tests are running with a compiled version of the code without the change to the `ps` command. If I do `make clean && make` and run the test again, it passes. There's something wrong with how the binaries are compiled for the tests. It seems to have nothing to do with your changes.

**Comment by vpodzime** [comment]
I tried splitting things up into the change of the code followed by the new tests in a separate commit, but that made no difference.

**Comment by tdlee** [comment]
Many thanks!

So in order to get things moving, who needs to do what next?  We seem to have both the 'triggering' aspect of my merge request, and the potentially bigger issue of a bug within the build-for-test process itself and which my merge request is exposing.

From my point of view I have another development which is waiting on this test, linked to CFE-1655 and CFE-3020.  See the comments from March/25 on CFE-1655.

**Comment by vpodzime** [comment]
> So in order to get things moving, who needs to do what next?

I need to find a way to make this pass tests.

**Comment by tdlee** [comment]
Thanks!  All the best with it.

**Comment by vpodzime** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2416/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2416/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2416/

**Comment by vpodzime** [comment]
So, after some cleanup, the test runs with agent using the changed `ps` options. And it fails in jenkins because the ssh session jenkins is using to access the build slave is not running with any `pts` terminal. And thus the test that checks that there is at least one process with a `pts` TTY is failing.

**Comment by tdlee** [comment]
Thanks.  Unlike most other internal-functional tests, this one can be sensitive to the environment.

Coming top-down:  the principle of "tty.cf" is "there should be a reasonable expectation of it passing in most reasonable environments".  So the guiding, top-down principle in the pattern-variable "dev_pattern" is "here is a pattern match that should encompass the bulk of common testing environments".

So it is reasonable that this pattern should encompass a typical developer on a typical platform, hence including Linux with one or more active "pts" sessions.  And you'll see that the comments and structure allow others (e.g. a Solaris developer) to make adjustments (then git-pull requests) to cover theirs.

It sounds as though this github/Jenkins environment may not have active terminal sessions.  (And that sounds understandable to me; fair enough.)  So what we need is some modification in there that can somehow encompass this github/Jenkins. If there is no obvious terminal session in this particular environment then we might have to use a fallback option of "in github/Jenkins skip this test".

So... just as it would be reasonable for a Solaris developer to flesh out the area around "dev_pattern" with something else and submit a git pull request, so I would be perfectly happy for you to do something similar, including if necessary that "skip if github/Jenkins".  As an outside user, I have no familiarity with these deep internals of github/Jenkins, so I would like to leave that to your expertise and wisdom!

Thanks.

**Comment by vpodzime** [comment]
@cf-bottom jenkins with exotics, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2423/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2423/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2423/

**Comment by vpodzime** [comment]
> Thanks. Unlike most other internal-functional tests, this one can be sensitive to the environment.

Yes, but I don't think it's too bad. We just need to make it pass jenkins. I extended the regular expression to also match `tty` and `ttyS` TTYs which seem to be allocated in our environments.

**Comment by tdlee** [comment]
Yes, that sounds good.  I look forward to this progressing!

**Comment by tdlee** [comment]
Any progress, please?  Thanks.

**Comment by vpodzime** [comment]
> Any progress, please? Thanks.

Progress is that I now again have capacity to look at this again.

**Comment by vpodzime** [comment]
So it looks like on the failed exotic platforms the code fails to match processes by tty using the `.+` regex.

**Comment by tdlee** [comment]
Thanks.  This area is known to be highly OS-variable. And the exact details of the comparison (regex or equivalent) will therefore need to be adjusted on an almost platform-by-platform basis.  See my comment above, dated April/1, and the commit linked from there. Also my comment from May/3 and your of May/6.  Thanks!

Is there anything further you need me to provide?

**Comment by tdlee** [comment]
A few days ago I was prompted for the Contributor Licence Agreement, which I completed.  Could we get this pull request, now over six months old, moving forwards, please.  Thanks.

**Comment by vpodzime** [comment]
>  Could we get this pull request, now over six months old, moving forwards, please. Thanks.

The only way to get this moving forward now is to only enable the test on Linux and HP-UX. Is that OK with you? Debugging and fixing the code on AIX and Solaris is beyond the importance of this feature (it doesn't seem to be widely used, especially on the exotic platforms).

**Comment by tdlee** [comment]
Yes, that's fine.&nbsp; As far as I can remember (and my memory isn't what it used to be!) this aspect was completely untested.&nbsp; This pull-request begins to address that, not least by adding the skeleton framework into which other platforms (e.g. the AIX and Solaris you mention) can be added over time.&nbsp; And it fleshes out (i.e. validates) that skeleton for the major platform of Linux.&nbsp; Some testing is better than none.

The background to this, by the way, was our bug report CFE-3020, which led to me consider implementing CFE-1655 (process handling via "/proc" where available) which in turn led me to realise that the existing "ps"-related testing was inadequate, hence my opening this CFE-3025.

So, yes please, on the basis that "incomplete testing is preferable to no testing", I would be keen to get this applied!

Thanks.

-- David Lee&nbsp;



On Mon, 14 Oct 2019 02:59:15 -0700, Vratislav Podzimek &lt;notifications@github.com&gt; wrote:
&nbsp;


Could we get this pull request, now over six months old, moving forwards, please. Thanks.


The only way to get this moving forward now is to only enable the test on Linux and HP-UX. Is that OK with you? Debugging and fixing the code on AIX and Solaris is beyond the importance of this feature (it doesn't seem to be widely used, especially on the exotic platforms).

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/3003/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/3003/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-3003/

**Comment by vpodzime** [comment]
Seems to be passing now. The failure on AIX 5.3 is unrelated to the changes being done in this PR.

**Comment by tdlee** [comment]
Thank you!  Much appreciated.
